### ISSUE TITLE
DQN with state vectors not frames

### ISSUE THREAD

**Issue by ThGravo** [astooke/rlpyt#67]
I wanted to start from example_1 and apply it to my specific environment. The state is a vector of 5001 floats. For some reason, this is interpreted as a sequence of frames:
```
Frame-based buffer using 5001-frame sequences.
```

How do I make sure it is doing the right thing?

**Comment by astooke** [comment]
Hi!  Yes take a look inside the DQN algorithm where it initializes the replay buffer using one of the `Frame` buffers.  Instead you could simply use the `UniformReplayBuffer` from `rlpyt/replays/non_sequence/uniform.py`, for example.  Does this work?

**Comment by ThGravo** [comment]
I tried but get this:
```
Traceback (most recent call last):
  File "RL/example_1.py", line 92, in <module>
    build_and_train(args, device, run_ID=0)
  File "RL/example_1.py", line 74, in build_and_train
    runner.train()
  File ".../rlpyt/rlpyt/runners/minibatch_rl.py", line 238, in train
    opt_info = self.algo.optimize_agent(itr, samples)
  File ".../rlpyt/rlpyt/algos/dqn/dqn.py", line 145, in optimize_agent
    samples_from_replay = self.replay_buffer.sample_batch(self.batch_size)
  File ".../rlpyt/rlpyt/replays/non_sequence/uniform.py", line 11, in sample_batch
    T_idxs, B_idxs = self.sample_idxs(batch_B)
  File ".../rlpyt/rlpyt/replays/non_sequence/uniform.py", line 18, in sample_idxs
    T_idxs = np.random.randint(low=low, high=high, size=(batch_B,))
  File "mtrand.pyx", line 630, in numpy.random.mtrand.RandomState.randint
  File "bounded_integers.pyx", line 1228, in numpy.random.bounded_integers._rand_int64
ValueError: low >= high
```

**Comment by astooke** [comment]
Hmm, could you report out your observation shape, and the `replay_buffer.samples.observation.shape`, `replay_buffer.off_backward`, `replay_buffer.off_forward`, and `replay_buffer.t` and `replay_buffer._buffer_full`   (basically everything about the observation shape and what goes into computing `low` and `high`)?

**Comment by astooke** [comment]
Any news or questions here?  Did you get it working??

**Comment by codelast** [comment]
@astooke I did what you recommended in my own customized environment(my state is also a vector) and it solved the "low >= high" problem.

**Comment by astooke** [comment]
Seems like this solved it, but if not, please reopen.  :)

**Comment by kargarisaac** [comment]
I'm trying to use rlpyt with my custom env with a non-image input state. For that, I first want to test it on a simple env, like CartPole-v0. And I use DQN and DqnAgent. But I get this error:
`
Traceback (most recent call last):
  File "/home/isaac/codes/dd-zero/rlpyt/examples/example_9.py", line 178, in <module>
    resume_chkpnt=None,
  File "/home/isaac/codes/dd-zero/rlpyt/examples/example_9.py", line 125, in build_and_train
    runner.train()
  File "/home/isaac/codes/dd-zero/rlpyt/rlpyt/runners/minibatch_rl.py", line 252, in train
    n_itr = self.startup()
  File "/home/isaac/codes/dd-zero/rlpyt/rlpyt/runners/minibatch_rl.py", line 81, in startup
    world_size=world_size,
  File "/home/isaac/codes/dd-zero/rlpyt/rlpyt/samplers/serial/sampler.py", line 47, in initialize
    global_B=global_B, env_ranks=env_ranks)
  File "/home/isaac/codes/dd-zero/rlpyt/rlpyt/agents/dqn/dqn_agent.py", line 37, in initialize
    global_B=global_B, env_ranks=env_ranks)
  File "/home/isaac/codes/dd-zero/rlpyt/rlpyt/agents/base.py", line 91, in initialize
    **self.model_kwargs)
TypeError: 'NoneType' object is not callable
`

The code is:
`
def build_and_train(run_ID=0, cuda_idx=None, resume_chkpnt=None):
    env_id = 'CartPole-v0'

    sampler = SerialSampler(
        EnvCls=gym_make,
        env_kwargs=dict(id=env_id),
        eval_env_kwargs=dict(id=env_id), 
        # env_kwargs=dict(id='CartPole-v0'),
        # eval_env_kwargs=dict(id='CartPole-v0'),
        batch_T=4,  # One time-step per sampler iteration.
        batch_B=8,  # One environment (i.e. sampler Batch dimension).
        max_decorrelation_steps=100,
        eval_n_envs=10,
        eval_max_steps=int(50e3),
        eval_max_trajectories=50,
    )

    algo = DQN(
        min_steps_learn=int(1e2),
        replay_size=int(1e4),
        replay_ratio=32,  # data_consumption / data_generation.
        learning_rate=0.01,
        double_dqn=True,
        ReplayBufferCls=UniformReplayBuffer,  # Leave None to select by above options.
    )

    agent = DqnAgent()

    runner = MinibatchRl(
        algo=algo,
        agent=agent,
        sampler=sampler,
        n_steps=1e6,
        log_interval_steps=1e3,
        affinity=dict(cuda_idx=cuda_idx, workers_cpus=[0,1,2,3,4,5,6])
    )

    config = dict(env_id=env_id)
    algo_name = 'dqn_'
    name = algo_name + env_id
    log_dir = algo_name + "cartpole"

    with logger_context(log_dir, run_ID, name, config, snapshot_mode='last'):
        runner.train()
`

But the ModelCls is None in DqnAgent and that's the reason for the error, I think. So I wrote an agent and model like this and used it instead of DqnAgent:
`
class CustomMixin:

    def make_env_to_model_kwargs(self, env_spaces):
        return dict(observation_shape=env_spaces.observation.shape,
                    output_size=env_spaces.action.n)
`
`
from rlpyt.agents.dqn.dqn_agent import DqnAgent
from rlpyt.utils.buffer import buffer_to
from rlpyt.agents.base import AgentStep
from rlpyt.utils.collections import namedarraytuple

import torch
AgentInfo = namedarraytuple("AgentInfo", "q")

class CustomDqnAgent(CustomMixin, DqnAgent):
    def __init__(self, ModelCls=DCustomDqnModel, **kwargs):
        super().__init__(ModelCls=ModelCls, **kwargs)
`

`
import torch

from rlpyt.utils.tensor import infer_leading_dims, restore_leading_dims
from rlpyt.models.mlp import MlpModel
from rlpyt.models.dqn.dueling import DuelingHeadModel

class CustomDqnModel(torch.nn.Module):

    def __init__(
            self,
            observation_shape,
            output_size,
            fc_sizes=128,
            dueling=False,
        ):
        """Instantiates the neural network according to arguments; network defaults
        stored within this method."""
        super().__init__()
        self.dueling = dueling
        self._obs_ndim = len(observation_shape)
        input_shape = observation_shape[0]
        self.base_net = torch.nn.Sequential(
            torch.nn.Linear(input_shape, fc_sizes),
            torch.nn.ReLU(),
            torch.nn.Linear(fc_sizes, fc_sizes),
            torch.nn.ReLU(),
            torch.nn.Linear(fc_sizes, output_size),
        )

    def forward(self, observation, prev_action, prev_reward):
        
        observation = observation.type(torch.float)
        lead_dim, T, B, obs_shape = infer_leading_dims(observation, self._obs_ndim)

        obs = observation.view(T * B, -1)
        q = self.base_net(obs)
        q = restore_leading_dims(q, lead_dim, T, B)
        return q
`

It runs fine, but It doesn't learn. 

![image](https://user-images.githubusercontent.com/24753756/77906414-993d2600-7290-11ea-9c61-efec91112dc3.png)

The plot is similar even after 20,000,000 steps. I checked the code several times and tested different configs for several days. 

Do you have any idea to solve this problem?

Thank you

**Comment by astooke** [comment]
@kargarisaac The problem might be if cartpole expects continuous actions, but the DQN algorithm is for discrete actions only.  Try DDPG or TD3 or SAC? (Also, this looks like a separate question and probably will be missed since it's at the end of a closed issue, recommend making a new issue next time.)

**Comment by kargarisaac** [comment]
Sry. I thought this issue is related to input vector states and asked here. I will open a new issue.

**Comment by seann999** [comment]
> @astooke I did what you recommended in my own customized environment(my state is also a vector) and it solved the "low >= high" problem.

How was this fixed
I'm stuck on the same problem

**Comment by seann999** [comment]
Okay I fixed it
I had to realize that using the ReplayBufferCls parameter in dqn.py did nothing, as it was an unused variable...

**Comment by astooke** [comment]
> I had to realize that using the ReplayBufferCls parameter in dqn.py did nothing, as it was an unused variable...

Oh! You're right!  That's embarrassing haha sorry about that...fixed in 85d4e01  :)  With a warning that it's totally up to the user to make sure the provided replay buffer class has all the right behaviors.
